Server tls-ekm negotiation

reynir opened the pull request:
This implements tls-ekm negotiation in the server. With this change I am able to reproduce #181 in the miragevpn-server-notun both with OpenVPN client and our Miragevpn client **when using tls 1.3**. When using tls 1.2 it works just fine with both clients.

reynir commented on src/miragevpn.mli:
```diff
@@ -217,11 +217,6 @@ module Config : sig
   val eq : eq
   (** [eq] is an implementation of [cmp] for use with [{!equal} cmp t t2] *)
 
-  val client_generate_connect_options :
-    t -> (string, [> `Msg of string ]) result
-  (** Exports the excerpts from the client configuration sent to the server
-      when the client initially connects. *)
-
```
This function seems very internal to me and I don't know why it was exposed in the first place - so let's not expose it anymore!

reynir commented on src/config.ml:
```diff
@@ -567,6 +586,11 @@ module Conf_map = struct
           | `Tcp (Some `Server) -> "-server"
           | `Tcp None | `Udp -> "")
+    | Protocol_flags, flags ->
+        (* XXX: We assume [flags] is always non-empty *)
+        p () "protocol-flags %a"
+          Fmt.(list ~sep:(any " ") (of_to_string Protocol_flag.to_string))
+          flags
```
There are no guarantees flags is not empty. However, a trailing space is not the end of the world...

reynir commented on src/config.ml:
```diff
@@ -202,6 +216,7 @@ module Conf_map = struct
           * [ `Udp | `Tcp of [ `Server | `Client ] option ])
           k
       (* see socket.c:static const struct proto_names proto_names[] *)
+    | Protocol_flags : Protocol_flag.t list k
```
I considered using a set, but a list is simpler and there's at most half a dozen different flags at the moment.

reynir commented on src/engine.ml:
```diff
@@ -558,7 +571,34 @@ let kex_server config session (my_key_material : my_key_material) tls data =
+    int_of_string_opt v
+  in
   let config = Config.add Cipher (cipher :> Config.cipher) config in
+  let config =
+    let supports_tls_ekm =
+      Option.fold iv_proto ~none:false
+        ~some:Packet.Iv_proto.(contains Tls_key_export)
+    in
+    let config =
+      if
+        supports_tls_ekm
+        && Option.fold iv_proto ~none:false
+             ~some:Packet.Iv_proto.(contains Use_cc_exit_notify)
+      then Config.add_protocol_flag `Tls_ekm config
+      else config
+    in
+    if supports_tls_ekm then Config.add Key_derivation `Tls_ekm config
+    else config
+  in
```
After testing it seems OpenVPN **only** sends `protocol-flags tls-ekm` and no `key-derivation tls-ekm` if `Use_cc_exit_notify` bit is set. I guess it could make the code slightly simpler only to add either.

reynir commented:
I patched OpenVPN to print whether tls-ekm or prf key derivation is used, and with TLS 1.3 tls-ekm is used - but somehow AEAD decrypt fails with tls 1.3.

hannesm commented:
Thanks for your work. To summarize:
- with tls-ekm and TLS 1.3, our server doesn't work with our client neither our server
- in #181 our client doesn't work with TLS 1.3 and ekm with an openvpn server

The "but somehow AEAD decrypt fails" is only with tls-ekm AFAIU.

Your other changes in here, would you mind to submit them as separate PR so we can merge them, and keep this for working/debugging on tls-ekm?

reynir commented:
I cleaned up the history including undoing sending `Tls_key_export` in the client. I also added a commit which effectively disables tls-ekm in the server while keeping the logic in a similar manner to how we did it in the client.

So I think this can be merged. Otherwise, I'm unsure what it is you want separated.

hannesm commented on src/config.ml:
```diff
@@ -1664,6 +1709,7 @@ let a_config_entry : line A.t =
          a_dhcp_option;
          a_proto;
          a_proto_force;
+         a_protocol_flags <?> "protocol-flags";
```
from what I remember, shouldn't we put the most eager at front -- i.e. don't let `a_proto` consume the "proto", but put the `a_protocol_flags` further above so it can try to consume the "protocol-flags"?

hannesm commented on src/config.ml:
```diff
@@ -2371,8 +2371,6 @@ let client_generate_connect_options t =
   (* Ok "V4,dev-type tun,link-mtu 1560,tun-mtu 1500,proto TCPv4_CLIENT,\
      keydir 1,cipher AES-256-CBC,auth SHA1,keysize 256,tls-auth,\
      key-method 2" *)
-  let open Result.Syntax in
-  let* () = Conf_map.is_valid_client_config t in
```
sounds fine to me.

hannesm reviewed (approved):
looks fine to me, apart from the minor comment about the order in the parser.

reynir commented on src/config.ml:
```diff
@@ -890,6 +919,22 @@ let a_proto_force =
   *> choice [ string "tcp" *> return `Tcp; string "udp" *> return `Udp ]
   >>| fun prot -> `Proto_force prot
 
+let a_protocol_flags =
+  string "protocol-flags" *> a_whitespace *> commit
+  *> sep_by a_whitespace a_single_param
```
Here we can parse an empty `protocol-flags` (which is fine but probably not used much). However, the empty `protocol-flags` requires at least one space or comment at the end to parse successfully. This is maybe not a big deal, but since the server may send such a protocol-flags it's unfortunate to error on that.

I'm not sure how to fix that.

reynir replied:
Figured it out. Tested manually with Miragevpn.Config.{parse,pp} in a top level.

hannesm commented:
Thanks for your replies and additional work.
